### PR TITLE
Add more docs on existing dynamic tool output features.

### PR DESCRIPTION
**Pull request by jmchilton**
Document the mysterious galaxy.json. 

xref https://github.com/galaxyproject/galaxy/pull/2697

### CODE REVIEW

File: docs/_writing_tool_metadata.rst [hunk 1]
```diff
@@ -0,0 +1,52 @@
+Tool Provided Metadata
+==============================
+
+This stub of a section provides some initial documentation on tool provided
+metadata. Galaxy allows any of datasets to be discovered after a tool has been
+executed and allows tool's to specify metadata for these datasets. Whenever
```
Comment (nsoranzo) [inline docs/_writing_tool_metadata.rst]: s/tool's/tools/

File: docs/_writing_tool_metadata.rst [hunk 2]
```diff
@@ -0,0 +1,52 @@
+Tool Provided Metadata
+==============================
+
+This stub of a section provides some initial documentation on tool provided
+metadata. Galaxy allows any of datasets to be discovered after a tool has been
```
Comment (nsoranzo) [inline docs/_writing_tool_metadata.rst]: s/any of datasets/any dataset/

sounds better to my non--native speaker ear.

File: docs/_writing_tool_metadata.rst [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+executed and allows tool's to specify metadata for these datasets. Whenever
+possible, Galaxy's datatypes and more structured outputs should
+be used to collect metadata.
+
+If arbitrary an number of outputs is needed but no special metadata must be set,
```
Comment (nsoranzo) [inline docs/_writing_tool_metadata.rst]: s/arbitrary an/an arbitrary/

File: docs/_writing_tool_metadata.rst [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+More information on this can be found on the `Galaxy wiki <https://wiki.galaxyproject.org/Admin/Tools/Multiple%20Output%20Files#Number_of_Output_datasets_cannot_be_determined_until_tool_run>`__.
+
+The file name patterns described in the above link are nice because they don't
+need special instrumenting in the tool wrapper to adapt to Galaxy in general and
+can adapt to many existing application's output. When more metadata must be 
+supplied or when implementing a custom tool wrapper anyway - it may be beneficial
+to build a manifest file.
+
+A tool may also produce a file called ``galaxy.json`` during execution. If 
+upon a job's completion this file is populated Galaxy will expect to find metadata
```
Comment (nsoranzo) [inline docs/_writing_tool_metadata.rst]: s/populated/populated,/

File: docs/_writing_tool_metadata.rst (hunk 2 above)
Comment (peterjc) [inline docs/_writing_tool_metadata.rst]: I agree

File: docs/_writing_tool_metadata.rst [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+Examples of tools using ``new_primary_dataset`` entries:
+
+- `tool_provided_metadata_2.xml <https://github.com/jmchilton/galaxy/blob/2909e74642180bd818019ebdcb62e62f12e56e69/test/functional/tools/tool_provided_metadata_2.xml>`__ demonstrating using the simpler attributes described here.
+- `tool_provided_metadata_3.xml <https://github.com/jmchilton/galaxy/blob/2909e74642180bd818019ebdcb62e62f12e56e69/test/functional/tools/tool_provided_metadata_3.xml>`__ demonstrates overridding datatype specified metadata.
+
+The ``type`` of an entry may also be ``dataset``. In this case the metadata 
+descriptions describe an explicit output (one with its own corresponding ``output``
+element definition). In this case, an entry called ``dataset`` should appear in
+the dictionary (in lieu of ``filename`` above) and should be the databaes id of the
```
Comment (nsoranzo) [inline docs/_writing_tool_metadata.rst]: s/databaes/database/

File: docs/writing_how_do_i.rst [hunk 6]
```diff
@@ -116,6 +116,21 @@ would check if the input is of type ``fastq`` or any derivative types such as
 - `Pull Request 457 <https://bitbucket.org/galaxy/galaxy-central/pull-request/457/allow-cheetah-tool-templates-to-reason/diff>`__
 
 -------------------------------------------------
+\.\.\. handle arbitrary output data formats?
+-------------------------------------------------
+
+If the output format of a tool's output cannot isn't known ahead of time,
```
Comment (nsoranzo) [inline docs/writing_how_do_i.rst]: s/cannot isn't known/is not known/

File: docs/writing_how_do_i.rst [hunk 7]
```diff
@@ -116,6 +116,21 @@ would check if the input is of type ``fastq`` or any derivative types such as
...
+Galaxy can be instructed to "sniff" the output and determine the data type
+using the same method used for uploads. Adding the ``auto_format="true"`
```
Comment (nsoranzo) [inline docs/writing_how_do_i.rst]: Missing "`" at the end of this line.

**Comment by jmchilton** [comment]
Thanks for the detailed review @nsoranzo - I fixed up things with 983d4aa I believe.

**Pull request merged**